examples/Example9/example9.cuh

```diff
@@ -70,6 +78,17 @@ public:
   __host__ __device__ RanluxppDoubleEngine(RanluxppDouble *engine)
       : G4HepEmRandomEngine(/*object=*/engine, &FlatWrapper, &FlatArrayWrapper)
   {
+#ifdef __CUDA_ARCH__
+    // This is a hack: The compiler cannot see that we're going to call the
+    // functions through their pointers, so it underestimates the number of
+    // required registers. By including calls to the (non-inlinable) functions
+    // we force the compiler to account for the register usage, even if this
+    // particular set of calls are not executed at runtime.
+    if (Zero) {
+      FlatWrapper(engine);
+      FlatArrayWrapper(engine, 0, nullptr);
+    }
```

Comment:
OK, if I understand correctly the call by pointer happens in G4HepEM I where the wrapper is passed and you need to show the compiler what it's calling into, and `RanluxppDouble::Rndm` uses quite a few registers. I wonder why it doesn't optimize out the if (Zero), but if it works...

Reply:
Yes, and not all functions from G4HepEm are inlined into `gammas.cu` yet (I need to make some changes there to make that possible).

The reason that the compiler cannot optimize out `if (Zero)` is that `Zero` is a global variable that could change its value. `__constant__` just means that its value doesn't change while a kernel is running, but there's an API function to change it from the host (which we use for setting up the data structures used by G4HepEm).